docs/design.md

```diff
@@ -0,0 +1,326 @@
+TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
+
+#### TLS 1.3
+
+TLS 1.3 offers enhanced security over its predecessor, all its supported cipher suites can be negotiated.
+
+### Authorisation
+
+For authorisation, a simple all-or-nothing mechanism will be used. The server can be started with a list of OS users that are allowed to make requests.
+
+X.509 certificates issued to users will have information about the user name encoded in them in the Common Name subject field. The server will then use this information to decide if a requester has been given permission to make requests.
+
+For example, using the snippet below the users `hashi`, `mozart` and `simba` would be able to make requests to the GRPC API.
+
+```
+./worker --allowedUsers=hashi,mozart,simba
+```
+
+## Worker Library
+
+The service will include a worker library written in Go to manage and interact with Linux and Linux processes.
+Under the hood, this is what the GRPC server will use to fulfill requests.
+
+A sketch of what this library could look like is below:
+
+```go
+// Worker is a Linux process manager
+// It keeps a registry of the processes that it
+// been requested to manage in memory.
+type Worker struct {
+	defaultresourceLimits ResourceLimits
+	execRegistry          map[ID]*ProcessHandle
+}
+
+// ID is the internal ID that the worker uses to identify an Exec
+// Under the hood this might just be a UUID
+type ID string
+
+type ResourceLimits struct {
+	MaxMemoryBytes int64
+	// ...
+}
+
+// ProcessRequest represents a request to execute a Linux process in the worker
+type ProcessRequest struct {
+	// Command is the command executed to be executed together with its arguments
+	Command string
+
+	// ResourceLimits specifies the resources that the process will have access to
+	// These translate to cgroup interface files configuration.
+	ResourceLimits ResourceLimits
+
+	RequestedBy string
+}
+
+// StartProcess starts a new process and adds it to the worker
+// process registry. It does not wait for the process to terminate.
+func (w *Worker) StartProcess(req ProcessRequest) (ID, error) {
+	// ...
+}
+
+// StopProcess stops a process currently managed by the worker
+// Returns an error if there was any issues stopping the process
+// or the process does not exist in the worker registry.
+func (w *Worker) StopProcess(processId ID) error {
+	// ...
+}
+
+// ProcessInfo gives access to information about a
+// process.
+type ProcessInfo interface {
+	PID() int
+	StartedBy() string
+	State() string
+	StartedAt() time.Time
+	FinishedAt() time.Time
+}
+
+// GetProcessInfo gives access to the ProcessInfo interface
+// which allows querying for information about the process.
+func (w *Worker) GetProcessInfo(processId ID) ProcessInfo {
+	// ...
+}
+
+// StreamProcessOutput returns an instance of a Streamer that
+// can be used to stream the combined stdout and stderr of
+// a process managed by the worker
+func (w *Worker) StreamProcessOutput(processId ID) Streamer {
```

Comment:
Maybe:

```suggestion
func (w *Worker) StreamProcessOutput(processId ID) (outputChan <-chan ProcessOutputEntry, errChan <-chan error) {
```

and ditch the Streamer interface?


Additionally, you might a direct error response here for situations like "job not found", although it could be sent via errChan. It's probably worth documenting either way.